### ISSUE TITLE
Possibly incorrect narrowing of overloads through function return types

### ISSUE THREAD

**Issue by Sharp-Eyes** [microsoft/pyright#6548]
For example, we construct a nested decorator, where the inner decorator can take one of two signatures and returns that same signature:
```py
import typing

Foo = typing.Callable[[str], str]
Bar = typing.Callable[[int, str], str]

@typing.overload
def deco(x: str) -> typing.Callable[[Foo], Foo]:
    ...

@typing.overload
def deco(x: str) -> typing.Callable[[Bar], Bar]:
    ...

def deco(x: str) -> typing.Callable[[Foo], Foo]:
    ...

@deco("a")
def foo(x: str) -> str:
    ...

@deco("b")   # ERROR: Function accepts too few positional parameters; expected 2 but received 1
def bar(num: int, x: str) -> str:
    ...
```
It does not appear that pyright takes the second overload into account at all. switching the overloads around (i.e. the `Bar`-overload first and the `Foo`-overload second) raises `"int" is incompatible with "str"` on the `foo` function instead.

This makes me believe overloads may only care about the function arguments and disregard the return type. I could understand this being intentional, but it'd be very nice to have for situations like this, so I decided to file this report.

## Workarounds

I am aware that something similar could be achieved through e.g.
```py
Foo = typing.Callable[[str], str]
Bar = typing.Callable[[int, str], str]
FooBarT = typing.TypeVar("FooBarT", Foo, Bar)


def deco(a: str) -> typing.Callable[[FooBarT], FooBarT]:
    ...
```
but (to my knowledge) I am unable to do this in my actual usecase due to `Foo` and `Bar` being generic themselves.

## Pyright Version
Pyright 1.1.337
Pylance v2023.11.13 (Pre-Release)

**Comment by erictraut** [comment]
Pyright is doing the right thing here. Your overload definitions are invalid because the second one completely overlaps the first one. Overloads are matched based on the input signature, not the return type.

An `overload` isn't the right tool for the job. What you're looking for is a type variable with an upper bound.

Try this instead:

```python
from typing import Any, Callable, TypeVar

T = TypeVar("T", bound=Callable[..., Any])

def deco(x: str) -> Callable[[T], T]:
    ...
```

Or if you are using Python 3.12, you can use the new type parameter syntax:

```python
from typing import Any, Callable

def deco[T: Callable[..., Any]](x: str) -> Callable[[T], T]:
    ...
```
